### PR TITLE
Feature/astral inventory gui

### LINKED ISSUE #79: The walls of the inner realm can be blown up.
Astral Version: astral-1.4.2
Patchouli Version: Patchouli-1.15.2-1.2-34
Forge Version: 1.15.2-forge-31.2.8

The player can blow up the walls of the inner realm and venture outside.
I've tested this with TNT, TNT minecarts, and end crystals, as the player can get those in their inner realm in survival by using an offering brazier.

### LINKED ISSUE #80: It is possible to break the walls of the inner realm using pistons.
Astral Version: astral-1.4.2
Patchouli Version: Patchouli-1.15.2-1.2-34
Forge Version: 1.15.2-forge-31.2.8

Ego membrane blocks break when something is pushed into them with a piston.

### LINKED ISSUE #92: Missing Tags
`Minecraft v1.15.2`
`Forge v31.2.31`
`Astral v1.4.15`

Astral leaves, logs, stripped logs, woods, planks, mushrooms, flowers, and saplings are missing their respective forge or Minecraft block and item tags, which would allow them to work in recipes.

Also, is Etheric Powder supposed to be placing buttons?

### LINKED ISSUE #93: Saplings grow with only 2 logs
`Minecraft v1.15.2`
`Forge v31.2.31`
`Astral v1.4.15`

When the astral sapling is grown, only one or two logs are spawned at the bottom, the rest are leaves.

### LINKED ISSUE #94: Add Ethereal Wood block and item
Make sure it can be stripped into Stripped Ethereal Wood

### PR DESCRIPTION
**Pull request by Alan19**
- Crystal Spiders drop crystal chitin
- Merge Astral and Inner Realm inventories, existing Inner Realm inventories will be dropped when entering the inner realm
- Non astral items get dropped when leaving inner realm
- Add a button that lets you check you Psychic Inventory
- Prevent pistons from interacting with ethereal blocks

### CODE REVIEW

File: src/main/java/com/alan19/astral/api/innerrealmteleporter/InnerRealmTeleporterCapability.java [hunk 1]
```diff
@@ -73,10 +78,29 @@ public void teleport(PlayerEntity player) {
             }
             if (!innerRealmWorld.isRemote()) {
                 TeleportationTools.performTeleport((ServerPlayerEntity) player, innerRealm, new BlockPos(playerSpawn.getX(), playerSpawn.getY(), playerSpawn.getZ()), player.getTeleportDirection());
+                dropInnerRealmItems(player, AstralAPI.getOverworldPsychicInventory((ServerWorld) innerRealmWorld).orElseGet(PsychicInventory::new).getInventoryOfPlayer(player.getUniqueID()));
```
Comment (SkySom) [inline src/main/java/com/alan19/astral/api/innerrealmteleporter/InnerRealmTeleporterCapability.java]: Why are you creating a new Inventory?

File: src/main/java/com/alan19/astral/api/innerrealmteleporter/InnerRealmTeleporterCapability.java [hunk 2]
```diff
@@ -73,10 +78,29 @@ public void teleport(PlayerEntity player) {
...
 
+    /**
+     * Drops all items being sent to the Inner Realm
+     *
+     * @param playerEntity      The player who was teleproted to the Inner Realm
+     * @param inventoryOfPlayer The PsychicInventoryInstance of the player
+     */
+    private void dropInnerRealmItems(PlayerEntity playerEntity, PsychicInventoryInstance inventoryOfPlayer) {
+        World entityWorld = playerEntity.world;
+        IntStream.range(0, inventoryOfPlayer.getInnerRealmMain().getSlots())
```
Comment (SkySom) [inline src/main/java/com/alan19/astral/api/innerrealmteleporter/InnerRealmTeleporterCapability.java]: You're probably better off just using a normal loop here.

File: src/main/java/com/alan19/astral/client/gui/AstralContainers.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import net.minecraftforge.common.extensions.IForgeContainerType;
+import net.minecraftforge.eventbus.api.IEventBus;
+import net.minecraftforge.fml.RegistryObject;
+import net.minecraftforge.registries.DeferredRegister;
+import net.minecraftforge.registries.ForgeRegistries;
+
+public class AstralContainers {
+    private static final DeferredRegister<ContainerType<?>> CONTAINERS = new DeferredRegister<>(ForgeRegistries.CONTAINERS, Astral.MOD_ID);
+
+    public static final RegistryObject<ContainerType<AstralInventoryContainer>> ASTRAL_INVENTORY_CONTAINER = CONTAINERS.register("astral_inventory", () -> IForgeContainerType.create((windowId, inv, data) -> new AstralInventoryContainer(windowId, inv)));
```
Comment (SkySom) [inline src/main/java/com/alan19/astral/client/gui/AstralContainers.java:14]: If you're not using Data, you don't need a ForgeContainerType

File: src/main/java/com/alan19/astral/client/gui/AstralInventoryContainer.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    private final PlayerInventory playerInventory;
+    private PsychicInventoryInstance inventory = new PsychicInventoryInstance();
+
+
+    public AstralInventoryContainer(int id, PlayerInventory playerInventory) {
+        super(AstralContainers.ASTRAL_INVENTORY_CONTAINER.get(), id);
+        this.playerInventory = playerInventory;
+        final PlayerEntity player = playerInventory.player;
+        if (player instanceof ServerPlayerEntity && AstralAPI.getOverworldPsychicInventory(((ServerPlayerEntity) player).getServerWorld()).isPresent()) {
+            inventory = AstralAPI.getOverworldPsychicInventory(((ServerPlayerEntity) player).getServerWorld()).orElseGet(PsychicInventory::new).getInventoryOfPlayer(player.getUniqueID());
```
Comment (SkySom) [inline src/main/java/com/alan19/astral/client/gui/AstralInventoryContainer.java]: Why are you creating a new PsychiInventory here? over mapping to get the inventory and the orElse to just create a dummy inventory.

File: src/main/java/com/alan19/astral/client/gui/AstralInventoryContainer.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+        //Add inventory
+        for (int row = 0; row < 3; row++) {
+            for (int col = 0; col < 9; col++) {
+                this.addSlot(new SlotItemHandler(inventory.getAstralMainInventory(), col + row * 9 + 9, 8 + col * 18, 84 + row * 18));
+            }
+        }
+
+        //Add hotbar
+        IntStream.range(0, 9)
```
Comment (SkySom) [inline src/main/java/com/alan19/astral/client/gui/AstralInventoryContainer.java:65]: Overuse of IntStream IMO

File: src/main/java/com/alan19/astral/events/astraltravel/InnerRealmEvents.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+import net.minecraftforge.fml.common.Mod;
+
+import java.util.stream.IntStream;
+
+@Mod.EventBusSubscriber(modid = Astral.MOD_ID)
+public class InnerRealmEvents {
+    @SubscribeEvent
+    public static void cancelInnerRealmEntry(EntityTravelToDimensionEvent event) {
+        if (event.getDimension().getRegistryName() != null && event.getDimension().getRegistryName().equals(AstralDimensions.INNER_REALM) && (event.getEntity() instanceof LivingEntity && !((LivingEntity) event.getEntity()).isPotionActive(AstralEffects.ASTRAL_TRAVEL.get())) || !(event.getEntity() instanceof LivingEntity)) {
```
Comment (SkySom) [inline src/main/java/com/alan19/astral/events/astraltravel/InnerRealmEvents.java]: You can nest ifs.

**Review by SkySom — CHANGES_REQUESTED** [review]

File: src/main/java/com/alan19/astral/client/gui/AstralContainers.java (hunk 3 above)
Comment (Alan19) [reply src/main/java/com/alan19/astral/client/gui/AstralContainers.java:14]: I'm not sure what I should replace it with

Comment (SkySom) [reply src/main/java/com/alan19/astral/client/gui/AstralContainers.java:14]: ContainerType.IFactory#create(windowId, PlayerInventory)

Comment (Alan19) [reply src/main/java/com/alan19/astral/client/gui/AstralContainers.java:14]: Still not sure on how to implement that. Am I supposed to be using new to create a new ContainerType.IFactory?

**Pull request merged**